tests/lib/rules/no-unused-class-component-methods.js

```diff
@@ -0,0 +1,388 @@
+/**
+ * @fileoverview Prevent declaring unused methods of component class
+ * @author Paweł Nowak
+ */
+'use strict';
+
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const rule = require('../../../lib/rules/no-unused-class-component-methods');
+const RuleTester = require('eslint').RuleTester;
+
+const parserOptions = {
+  ecmaVersion: 2018,
+  sourceType: 'module',
+  ecmaFeatures: {
+    jsx: true
+  }
+};
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({parserOptions});
+ruleTester.run('no-unused-class-component-methods', rule, {
+  valid: [
+    {
+      code: `
+        class Foo extends React.Component {
+          handleClick() {}
+          render() {
+            return <button onClick={this.handleClick}>Text</button>;
+          }
+        }
+      `
+    },
+    {
+      code: `
+        class Foo extends React.Component {
+          action() {}
+          componentDidMount() {
+            this.action();
+          }
+          render() {
+            return null;
+          }
+        }
+      `
+    },
+    {
+      code: `
+        class Foo extends React.Component {
+          action() {}
+          componentDidMount() {
+            const action = this.action;
+            action();
+          }
+          render() {
+            return null;
+          }
+        }
+      `
+    },
+    {
+      code: `
+        class Foo extends React.Component {
+          getValue() {}
+          componentDidMount() {
+            const action = this.getValue();
+          }
+          render() {
+            return null;
+          }
+        }
+      `
+    },
+    {
+      code: `
+        class Foo extends React.Component {
+          handleClick = () => {}
+          render() {
+            return <button onClick={this.handleClick}>Button</button>;
+          }
+        }
+      `,
+      parser: 'babel-eslint'
+    },
+    {
+      code: `
+        class Foo extends React.Component {
+          renderContent() {}
+          render() {
+            return <div>{this.renderContent()}</div>;
+          }
+        }
+      `
+    },
+    {
+      code: `
+        class Foo extends React.Component {
+          renderContent() {}
+          render() {
+            return (
+              <div>
+                <div>{this.renderContent()}</div>;
+              </div>
+            );
+          }
+        }
+      `
+    },
+    {
+      code: `
+        class Foo extends React.Component {
+          property = {}
```

Comment:
Why is it valid?

Reply:
Could it also check unused property?

Reply:
That's a good idea; the rule should check both properties and methods.

Reply:
IMO it's out of basic scope. Rule name is `no-unused-class-component-methods`. I feel we are trying to solve all issues with that single rule and this can overcomplicate it quickly.